File: androidApp/src/main/java/io/redlink/more/app/android/observations/Polar/Polar360PpiObservation.kt
package io.redlink.more.app.android.observations.Polar

import android.Manifest
import android.content.Context
import android.content.pm.PackageManager
import android.os.Build
import androidx.core.app.ActivityCompat
import com.polar.sdk.api.PolarBleApi
import com.polar.sdk.api.model.PolarPpiData
import io.github.aakira.napier.Napier
import io.reactivex.rxjava3.disposables.Disposable
import io.reactivex.rxjava3.schedulers.Schedulers
import io.redlink.more.app.android.MoreApplication
import io.redlink.more.app.android.R
import io.redlink.more.app.android.extensions.stringResource
import io.redlink.more.app.android.observations.pauseObservation
import io.redlink.more.app.android.observations.showPermissionAlertDialog
import io.redlink.more.app.android.services.sensorsListener.BluetoothStateListener
import io.redlink.more.database.repository.MainRepository
import io.redlink.more.extensions.anyNameIn
import io.redlink.more.observations.Observation
import io.redlink.more.observations.observationTypes.Polar360PpiType
import io.redlink.more.scopes.Scope
import io.redlink.more.services.bluetooth.BluetoothStateManagement
import kotlinx.coroutines.Job

private val permissions =
    if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.S) {
        if (Build.VERSION.SDK_INT >= 34) {
            setOf(
                Manifest.permission.BLUETOOTH_SCAN,
                Manifest.permission.BLUETOOTH_CONNECT,
                Manifest.permission.BLUETOOTH_ADVERTISE,
                Manifest.permission.ACCESS_FINE_LOCATION,
                Manifest.permission.FOREGROUND_SERVICE_CONNECTED_DEVICE
            )
        } else {
            setOf(
                Manifest.permission.BLUETOOTH_SCAN,
                Manifest.permission.BLUETOOTH_CONNECT,
                Manifest.permission.ACCESS_FINE_LOCATION
            )
        }
    } else {
        setOf(
            Manifest.permission.BLUETOOTH,
            Manifest.permission.BLUETOOTH_ADMIN
        )
    }

class Polar360PpiObservation(repos: MainRepository) :
    Observation(repos, observationType = Polar360PpiType(permissions)) {

    private val deviceIdentifier = setOf("Polar")
    private var ppiDisposable: Disposable? = null
    private var offlineRecordingDisposable: Disposable? = null
    private var deviceConnectionListener: Job? = null
    private var offlineMode = false

    companion object {
        var recording_startTimestamp: Long? = null
        var recroding_endTimestamp: Long? = null

        fun padToOneHz(
            samples: List<ppi_data>,
            startNs: Long,
            endNs: Long
        ): List<ppi_data> {
            if (endNs <= 0 || samples.isEmpty()) return samples

            val step = 1_000_000_000L
            val sorted = samples.sortedBy { it.timestamp }

            // Window start: take the earlier of recording_startTimestamp and first sample,
            // floored to a whole second, so the full recording period is covered.
            val firstSampleTs = sorted.first().timestamp
            val rawStart = if (startNs > 0) minOf(startNs, firstSampleTs) else firstSampleTs
            val effectiveStart = (rawStart / step) * step

            if (effectiveStart >= endNs) return samples

            val maxDurationNs = 24L * 3600 * 1_000_000_000L
            if (endNs - effectiveStart > maxDurationNs) {
                Napier.w(tag = "Polar360PpiObservation") { "padToOneHz — derived duration exceeds 24 h cap; returning raw samples" }
                return samples
            }

            val result = mutableListOf<ppi_data>()
            var sampleIndex = 0
            var cursor = effectiveStart

            while (cursor < endNs) {
                val slotEnd = cursor + step
                // Advance past any samples that fall before this slot.
                while (sampleIndex < sorted.size && sorted[sampleIndex].timestamp < cursor) {
                    sampleIndex++
                }
                if (sampleIndex < sorted.size && sorted[sampleIndex].timestamp < slotEnd) {
                    val s = sorted[sampleIndex]
                    result.add(ppi_data(hr = s.hr, timestamp = cursor, ppiInMs = s.ppiInMs, ppiErrorEstimate = s.ppiErrorEstimate, skinContact = s.skinContact))
                    sampleIndex++
                    // Skip any additional samples within the same 1-second slot.
                    while (sampleIndex < sorted.size && sorted[sampleIndex].timestamp < slotEnd) {
                        sampleIndex++
                    }
                } else {
                    result.add(ppi_data(hr = -99, timestamp = cursor, ppiInMs = 0, ppiErrorEstimate = 65535, skinContact = false))
                }
                cursor = slotEnd
            }

            val validCount = result.count { it.ppiErrorEstimate != 65535 }
            Napier.d(tag = "Polar360PpiObservation") { "padToOneHz — total=${result.size}, valid=$validCount, corrupted=${result.size - validCount}" }
            return result
        }
    }

    override fun start(): Boolean {
        Napier.d(tag = "Polar360PpiObservation::start") { "Starting Polar 360 PPI (offline=$offlineMode)..." }
        if (!observerAccessible()) {
            showObservationErrorNotification(
                stringResource(R.string.observation_cannot_start),
                stringResource(R.string.observation_error)
            )
            return false
        }

        val device = Polar360Controller.findPolar360Device()
        if (device == null) {
            Napier.d(tag = "Polar360PpiObservation::start") { "No Polar 360 device connected" }
            showObservationErrorNotification(
                stringResource(R.string.observation_cannot_start),
                stringResource(R.string.observation_error)
            )
            return false
        }

        val deviceId = device.deviceId!!
        Polar360Controller.ensureReady(deviceId, offlineMode = offlineMode,
            onReady = {
                if (offlineMode) {
                    Polar360Controller.stopOfflineRecording(PolarBleApi.PolarDeviceDataType.PPI)
                        .subscribe(
                            { items ->
                                val processed = processPpiSamples(items.filterIsInstance<PolarPpiData.PolarPpiSample>())
                                if (processed.isNotEmpty()) {
                                    val padded = padToOneHz(
                                        samples = processed,
                                        startNs = recording_startTimestamp ?: 0L,
                                        endNs = recroding_endTimestamp ?: 0L
                                    )
                                    storeData(mapOf("polar360ppidata" to padded), -1) {}
                                }
                                offlineRecordingDisposable = Polar360Controller.startOfflineRecording(
                                    deviceId, PolarBleApi.PolarDeviceDataType.PPI
                                )
                            },
                            { error ->
                                Napier.e(tag = "Polar360PpiObservation") { "Failed to fetch pending offline data: ${error.message}" }
                                offlineRecordingDisposable = Polar360Controller.startOfflineRecording(
                                    deviceId, PolarBleApi.PolarDeviceDataType.PPI
                                )
                            }
                        )
                } else {
                    ppiDisposable = Polar360Controller.getPolarApi()
                        .startPpiStreaming(deviceId)
                        .subscribeOn(Schedulers.io())
                        .observeOn(Schedulers.io())
                        .subscribe(
                            { data: PolarPpiData ->
                                if (data.samples.isNotEmpty()) {
                                    val processed = processPpiSamples(data.samples)
                                    processed.forEach { storeData(it) }
                                }
                            },
                            { error ->
                                Napier.e(tag = "Polar360PpiObservation") { "PPI stream failed: ${error.message}" }
                                ppiDisposable = null
                            }
                        )
                }
            },
            onError = { error ->
                Napier.e(tag = "Polar360PpiObservation") { "Setup failed: ${error.message}" }
                showObservationErrorNotification(
                    stringResource(R.string.observation_cannot_start),
                    stringResource(R.string.observation_error)
                )
            }
        )

        deviceConnectionListener = listenToDeviceConnection()
        return true
    }

    override fun stop(onCompletion: () -> Unit) {
        deviceConnectionListener?.cancel()
        deviceConnectionListener = null
        if (offlineMode) {
            offlineRecordingDisposable?.dispose()
            offlineRecordingDisposable = null
            Polar360Controller.stopOfflineRecording(PolarBleApi.PolarDeviceDataType.PPI)
                .subscribe(
                    { items ->
                        val processed = processPpiSamples(items.filterIsInstance<PolarPpiData.PolarPpiSample>())
                        val padded = padToOneHz(
                            samples = processed,
                            //todo polar uses non traditional timestamp start date
                            startNs = recording_startTimestamp ?: 0L,
                            endNs = recroding_endTimestamp ?: 0L
                        )
                        storeData(mapOf("polar360ppidata" to padded), -1, onCompletion)
                    },
                    { error ->
                        Napier.e(tag = "Polar360PpiObservation") { "Failed to process offline data: ${error.message}" }
                        onCompletion()
                    }
                )
        } else {
            ppiDisposable?.dispose()
            ppiDisposable = null
            onCompletion()
        }
    }

    data class ppi_data(val hr: Int, val timestamp: Long, val ppiInMs: Int, val ppiErrorEstimate: Int , var skinContact: Boolean)

    fun processPpiSamples(samples: List<PolarPpiData.PolarPpiSample>?): List<ppi_data> {
        if (samples.isNullOrEmpty()) return emptyList()
        return samples.map {
            ppi_data(hr = it.hr, timestamp = it.timeStamp.toLong(), ppiInMs = it.ppi, ppiErrorEstimate = it.errorEstimate , skinContact =  it.skinContactStatus )
        }
    }

    override fun observerErrors(): Set<String> {
        val errors = mutableSetOf<String>()
        if (!hasPermissions(MoreApplication.appContext!!)) {
            errors.add("error_access_bluetooth")
            showPermissionAlertDialog()
        }
        if (!BluetoothStateListener.bluetoothEnabled.value) {
            errors.add("bluetooth_disabled")
        }
        if (!MoreApplication.shared!!.bluetoothController.observerDeviceAccessible(deviceIdentifier)) {
            errors.add("device_not_connected")
            errors.add(ERROR_DEVICE_NOT_CONNECTED)
        }
        return errors
    }

    override fun bleDevicesNeeded(): Set<String> = deviceIdentifier

    override fun ableToAutomaticallyStart(): Boolean = observerAccessible()

    override fun applyObservationConfig(settings: Map<String, Any>) {
        offlineMode = Polar360Controller.isOfflineRecordingMode(settings)
    }

    private fun hasPermissions(context: Context): Boolean {
        return permissions.all {
            ActivityCompat.checkSelfPermission(context, it) != PackageManager.PERMISSION_DENIED
        }
    }

    private fun listenToDeviceConnection(): Job {
        return Scope.launch {
            BluetoothStateManagement.connectedDevices.collect { devices ->
                if (!deviceIdentifier.anyNameIn(devices)) {
                    pauseObservation(Polar360PpiType(emptySet()))
                    Polar360Controller.onDeviceDisconnected()
                }
            }
        }.second
    }
}

File: iosApp/iosApp/Observations/Polar360PpiObservation.swift
//
//  Polar360PpiObservation.swift
//  iosApp
//

import Combine
import CoreBluetooth
import Foundation
import KMPNativeCoroutinesCombine
import PolarBleSdk
import RxSwift
import shared

class Polar360PpiObservation: Observation_ {

    private let deviceIdentificer: Set<String> = ["Polar", "360"]
    private let bleManager = BluetoothStateManagement.shared
    private let controller = Polar360Controller.shared

    private var ppiDisposable: Disposable?
    private var offlineRecordingDisposable: Disposable?
    private var deviceListener: AnyCancellable?
    private var offlineMode = false
    private var stopBackgroundSync: BackgroundSync?

    private static let notificationBackoffInterval: TimeInterval = 60
    private static var lastCannotStartNotificationDate: Date?
    
    static var recording_startTimestamp: UInt64? = nil
    static var recroding_endTimestamp: UInt64? = nil
    
    init(repos: MainRepository, sensorPermissions: Set<String>) {
        super.init(repos: repos, observationType: Polar360PpiType(sensorPermissions: sensorPermissions))
    }
   

    override func start() -> Bool {
        if !observerAccessible() {
            showCannotStartNotification()
            return false
        }

        guard let device = controller.findPolar360Device() else {
            showCannotStartNotification()
            return false
        }
        let deviceId = device.deviceId

        listenToDeviceConnection()

        controller.ensureReady(deviceId: deviceId, offlineMode: offlineMode, onReady: { [weak self] in
                guard let self else { return }
                if self.offlineMode {
                    self.controller.stopOfflineRecordingAndFetch(
                        dataType: .ppi,
                        onSuccess: { [weak self] items in
                            guard let self else { return }
                            let samples = items.compactMap { $0 as? Polar360Controller.ppi_data }
                            if !samples.isEmpty {
                                let padded = Polar360PpiObservation.padToOneHz(
                                    samples: samples,
                                    startNs: Polar360PpiObservation.recording_startTimestamp ?? 0,
                                    endNs: Polar360PpiObservation.recroding_endTimestamp ?? 0
                                )
                                let processed = padded.map { ["hr": $0.hr, "ppiInMs": $0.ppiInMs, "ppiErrorEstimate": $0.ppiErrorEstimate, "timestamp": $0.timestamp, "skinContact": $0.skinContact] as [String: Any] }
                                self.storeData(data: ["polar360ppidata": processed], timestamp: -1) {}
                            }
                            self.offlineRecordingDisposable = self.controller.startOfflineRecording(
                                deviceId: deviceId, dataType: .ppi
                            )
                        },
                        onError: { [weak self] error in
                            Napier.e("Polar360PpiObservation: Failed to fetch pending offline data: \(error)")
                            guard let self else { return }
                            self.offlineRecordingDisposable = self.controller.startOfflineRecording(
                                deviceId: deviceId, dataType: .ppi
                            )
                        }
                    )
                } else {
                    self.ppiDisposable = self.controller.getPolarApi()
                        .startPpiStreaming(deviceId)
                        .subscribe(on: MainScheduler.instance)
                        .observe(on: ConcurrentDispatchQueueScheduler(qos: .background))
                        .subscribe(
                            onNext: { [weak self] data in
                                guard let self else { return }
                                for sample in data.samples {
                                    self.storeData(
                                        data: [
                                            "hr": sample.hr,
                                            "ppiInMs": sample.ppInMs,
                                            "ppiErrorEstimate": sample.ppErrorEstimate,
                                            "timestamp": sample.timeStamp,
                                            "skinContact": sample.skinContactStatus == 1
                                        ],
                                        timestamp: -1
                                    ) { }
                                }
                            },
                            onError: { error in
                                Napier.e("Polar360 PPI stream failed: \(error)")
                            }
                        )
                }
            
        }, onError: { error in Napier.e("Polar360 PPI setup error: \(error)") })
        return true
    }

    override func stop(onCompletion: @escaping () -> Void) {
        deviceListener?.cancel()
        deviceListener = nil
        if offlineMode {
            offlineRecordingDisposable?.dispose()
            offlineRecordingDisposable = nil

            let finishBg: () -> Void
            if controller.appIsInBackground {
                let sync = BackgroundSync()
                stopBackgroundSync = sync
                finishBg = sync.begin(taskName: "Polar360PpiStopAndFetch")
            } else {
                finishBg = { [weak self] in self?.stopBackgroundSync = nil }
            }

            controller.stopOfflineRecordingAndFetch(
                dataType: .ppi,
                onSuccess: { [weak self] items in
                    guard let self else { onCompletion(); finishBg(); return }
                    let samples = items.compactMap { $0 as? Polar360Controller.ppi_data }
                    guard !samples.isEmpty else { onCompletion(); finishBg(); return }
                    let padded = Polar360PpiObservation.padToOneHz(
                        samples: samples,
                        startNs: Polar360PpiObservation.recording_startTimestamp ?? 0,
                        endNs: Polar360PpiObservation.recroding_endTimestamp ?? 0
                    )
                    let processed = padded.map { ["hr": $0.hr, "ppiInMs": $0.ppiInMs, "ppiErrorEstimate": $0.ppiErrorEstimate, "timestamp": $0.timestamp, "skinContact": $0.skinContact] as [String: Any] }
                    self.storeData(data: ["polar360ppidata": processed], timestamp: -1) {
                        onCompletion()
                        finishBg()
                    }
                },
                onError: { error in
                    Napier.e("Polar360PpiObservation: Failed to fetch offline data: \(error)")
                    onCompletion()
                    finishBg()
                }
            )
        } else {
            ppiDisposable?.dispose()
            ppiDisposable = nil
            onCompletion()
        }
    }

    /// Pads PPI samples to 1 Hz across [effectiveStart, endNs].
    /// Both startNs, endNs, and sample timestamps are in nanoseconds since 2000-01-01.
    /// Each 1-second slot gets the first sample whose timestamp falls in [cursor, cursor+1s),
    /// or a dummy (hr=-99, ppiErrorEstimate=UInt16.max) if no sample falls in that slot.
    static func padToOneHz(
        samples: [Polar360Controller.ppi_data],
        startNs: UInt64,
        endNs: UInt64
    ) -> [Polar360Controller.ppi_data] {
        guard endNs > 0, !samples.isEmpty else { return samples }

        let step: UInt64 = 1_000_000_000
        let sorted = samples.sorted { $0.timestamp < $1.timestamp }

        // Window start: take the earlier of recording_startTimestamp and first sample,
        // floored to a whole second, so the full recording period is covered.
        let firstSampleTs = sorted.first!.timestamp
        let rawStart = startNs > 0 ? min(startNs, firstSampleTs) : firstSampleTs
        let effectiveStart = (rawStart / step) * step

        guard effectiveStart < endNs else { return samples }

        // Safety cap: never generate more than 24 h of slots.
        let maxDurationNs: UInt64 = 24 * 3600 * 1_000_000_000
        guard endNs - effectiveStart <= maxDurationNs else {
            Napier.w("Polar360PpiObservation: padToOneHz — derived duration exceeds 24 h cap; returning raw samples")
            return samples
        }

        var result: [Polar360Controller.ppi_data] = []
        var sampleIndex = 0
        var cursor = effectiveStart

        while cursor < endNs {
            let slotEnd = cursor + step
            // Advance past any samples that fall before this slot.
            while sampleIndex < sorted.count && sorted[sampleIndex].timestamp < cursor {
                sampleIndex += 1
            }
            if sampleIndex < sorted.count && sorted[sampleIndex].timestamp < slotEnd {
                let s = sorted[sampleIndex]
                result.append(Polar360Controller.ppi_data(
                    hr: s.hr,
                    timestamp: cursor,
                    ppiInMs: s.ppiInMs,
                    ppiErrorEstimate: s.ppiErrorEstimate,
                    skinContact: s.skinContact ? 1 : 0
                ))
                sampleIndex += 1
                // Skip any additional samples within the same 1-second slot.
                while sampleIndex < sorted.count && sorted[sampleIndex].timestamp < slotEnd {
                    sampleIndex += 1
                }
            } else {
                result.append(Polar360Controller.ppi_data(
                    hr: -99,
                    timestamp: cursor,
                    ppiInMs: 0,
                    ppiErrorEstimate: .max,
                    skinContact: 0
                ))
            }
            cursor += step
        }

        let validCount = result.filter { $0.ppiErrorEstimate != .max }.count
        Napier.d("Polar360PpiObservation: padToOneHz — total=\(result.count), valid=\(validCount), corrupted=\(result.count - validCount)")
        return result
    }

    override func observerErrors() -> Set<String> {
        var errors: Set<String> = []
        if CBManager.authorization != .allowedAlways {
            errors.insert("Access to Bluetooth not granted")
            PermissionManager.openSensorPermissionDialog()
        }
        if !bleManager.bluetoothActiveValue {
            errors.insert("Bluetooth is not enabled")
        }
        if !AppDelegate.shared.bluetoothController.observerDeviceAccessible(bleDevices: deviceIdentificer) {
            errors.insert("No polar device connected")
            errors.insert(Observation_.companion.ERROR_DEVICE_NOT_CONNECTED)
        }
        return errors
    }

    override func applyObservationConfig(settings: Dictionary<String, Any>) {
        offlineMode = controller.isOfflineRecordingMode(config: settings)
    }

    override func bleDevicesNeeded() -> Set<String> { deviceIdentificer }

    override func ableToAutomaticallyStart() -> Bool { observerAccessible() }

    private func showCannotStartNotification() {
        let now = Date()
        if let last = Polar360PpiObservation.lastCannotStartNotificationDate,
           now.timeIntervalSince(last) < Polar360PpiObservation.notificationBackoffInterval {
            return
        }
        Polar360PpiObservation.lastCannotStartNotificationDate = now
        showObservationErrorNotification(
            notificationBody: "Cannot start PPI Observation! Please enable Bluetooth and connect devices.",
            fallbackTitle: "Observation Error"
        )
    }

    private func listenToDeviceConnection() {
        deviceListener = createPublisher(for: bleManager.connectedDevices)
            .removeDuplicates()
            .receive(on: DispatchQueue.main)
            .sink(receiveCompletion: { _ in }, receiveValue: { [weak self] devices in
                if let self, !self.deviceIdentificer.anyNameIn(items: devices) {
                    self.controller.onDeviceDisconnected()
                    self.deviceListener?.cancel()
                }
            })
    }
}
